// @morev/vue-transitions/nuxt/module.mjs
import { readFileSync, existsSync, unlinkSync, mkdirSync, writeFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { isArray, mergeObjects, isEmpty } from '@morev/utils';
import { defineNuxtModule, createResolver, isNuxt2, addComponentsDir } from '@nuxt/kit';

const __dirname = dirname(fileURLToPath(import.meta.url));
const COMPONENTS = ["TransitionExpand", "TransitionFade", "TransitionScale", "TransitionSlide"];
const BABEL_PLUGIN_NAME = "@babel/plugin-transform-logical-assignment-operators";
const SCOPE = "@morev";
const MODULE_NAME = `${SCOPE}/vue-transitions`;
const module = defineNuxtModule({
  meta: {
    name: `${MODULE_NAME}/nuxt`,
    configKey: "vueTransitions",
    compatibility: {
      nuxt: ">= 2.17.0 || >=3.5.0"
    }
  },
  defaults: {
    componentDefaultProps: {},
    defaultProps: {}
  },
  async setup(options, nuxt) {
    var _a, _b;
    const NODE_MODULES_PATH = __dirname.replace(new RegExp(`${SCOPE}.*`), "");
    const COMPONENTS_DIRECTORY = join(NODE_MODULES_PATH, ".vue-transitions");
    const resolver = createResolver(import.meta.url);
    (_a = nuxt.options).css ?? (_a.css = []);
    nuxt.options.css.push(`${MODULE_NAME}/styles`);
    if (isNuxt2()) {
      nuxt.options.build.transpile.push("@morev/utils", "ohash", MODULE_NAME);
      (_b = nuxt.options.build.babel).plugins ?? (_b.plugins = []);
      const doesBabelPluginExists = nuxt.options.build.babel.plugins.some((plugin) => {
        return isArray(plugin) ? plugin[0] === BABEL_PLUGIN_NAME : plugin === BABEL_PLUGIN_NAME;
      });
      !doesBabelPluginExists && nuxt.options.build.babel.plugins.push(BABEL_PLUGIN_NAME);
    }
    const templateContents = readFileSync(resolver.resolve("template.vue"), { encoding: "utf8" });
    try {
      existsSync(COMPONENTS_DIRECTORY) && unlinkSync(COMPONENTS_DIRECTORY);
    } catch {
    }
    mkdirSync(COMPONENTS_DIRECTORY, { recursive: true });
    COMPONENTS.forEach((componentName) => {
      const customProps = mergeObjects(
        options.defaultProps ?? {},
        options.componentDefaultProps?.[componentName]
      );
      const propsDeclaration = isEmpty(customProps) ? "$attrs" : JSON.stringify(customProps).replace(/}$/, ",...$$attrs}").replaceAll('"', "'");
      writeFileSync(
        join(COMPONENTS_DIRECTORY, `${componentName}.vue`),
        templateContents.replaceAll("<%= options.propsDeclaration %>", propsDeclaration).replaceAll("<%= options.listenersDeclaration %>", isNuxt2() ? ' v-on="$listeners"' : "").replaceAll("<%= options.componentName %>", componentName)
      );
    });
    addComponentsDir({
      path: COMPONENTS_DIRECTORY,
      pathPrefix: false,
      watch: false
    });
  }
});

export { module as default };
