// @morev/utils/dist/guards/index.js
import "../chunk-BT72POBR.js";
import {
  isSessionStorageAvailable
} from "../chunk-2HW42MFA.js";
import {
  isSymbol
} from "../chunk-TKM2FQTA.js";
import {
  isWeakMap
} from "../chunk-Y3T3BIRX.js";
import {
  isPlainObject
} from "../chunk-EWB5CQOX.js";
import {
  isPrimitive
} from "../chunk-JL4YVFFX.js";
import {
  isNullish
} from "../chunk-Z5IVBZLV.js";
import {
  isUndefined
} from "../chunk-F45GO7K2.js";
import {
  isNull
} from "../chunk-TK4YRMRC.js";
import {
  isPromise
} from "../chunk-SPK7OES6.js";
import {
  isRegExp
} from "../chunk-5HZ32HEL.js";
import {
  isFormData
} from "../chunk-ON2BNX4R.js";
import {
  isFunction
} from "../chunk-CQJCSTBU.js";
import {
  isHex
} from "../chunk-GEEDF44I.js";
import {
  isInteger
} from "../chunk-4W3EM56D.js";
import {
  isLocalStorageAvailable
} from "../chunk-MKT6NLMZ.js";
import {
  isMap
} from "../chunk-EXCPRRER.js";
import {
  isNode
} from "../chunk-M3TQVFXR.js";
import {
  isDate
} from "../chunk-4GTX7FBA.js";
import {
  isEmail
} from "../chunk-3SNQM74H.js";
import {
  isEmpty
} from "../chunk-ED5RUP25.js";
import {
  isObject
} from "../chunk-XUWL4VKZ.js";
import {
  isIterable
} from "../chunk-HJKYC4E4.js";
import {
  isError
} from "../chunk-WE7QC4OC.js";
import {
  isFile
} from "../chunk-TZZM7RNI.js";
import {
  isFloat
} from "../chunk-MV2MTLNA.js";
import {
  isNumeric
} from "../chunk-OBG63REY.js";
import {
  isArray
} from "../chunk-UDVSPJHY.js";
import {
  isBlob
} from "../chunk-UDW3FTZT.js";
import {
  isBoolean
} from "../chunk-42JGV6MK.js";
import {
  isElement
} from "../chunk-FMEA3XAZ.js";
import {
  isString
} from "../chunk-RUB2ELNH.js";
export {
  isArray,
  isBlob,
  isBoolean,
  isDate,
  isElement,
  isEmail,
  isEmpty,
  isError,
  isFile,
  isFloat,
  isFormData,
  isFunction,
  isHex,
  isInteger,
  isIterable,
  isLocalStorageAvailable,
  isMap,
  isNode,
  isNull,
  isNullish,
  isNumeric,
  isObject,
  isPlainObject,
  isPrimitive,
  isPromise,
  isRegExp,
  isSessionStorageAvailable,
  isString,
  isSymbol,
  isUndefined,
  isWeakMap
};
