// @morev/utils/dist/environment/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/environment/index.ts
var environment_exports = {};
__export(environment_exports, {
  isClient: () => isClient,
  isDev: () => isDev,
  isFirefox: () => isFirefox,
  isInternetExplorer: () => isInternetExplorer,
  isLighthouse: () => isLighthouse,
  isMobile: () => isMobile,
  isMotionless: () => isMotionless,
  isOSX: () => isOSX,
  isProd: () => isProd,
  isServer: () => isServer,
  isTest: () => isTest,
  isTouch: () => isTouch,
  supportsPassive: () => supportsPassive
});
module.exports = __toCommonJS(environment_exports);

// src/environment/is-client/is-client.ts
var isClient = () => typeof window !== "undefined";

// src/environment/is-dev/is-dev.ts
var isDev = () => {
  if (typeof process === "undefined")
    return null;
  return process.env.NODE_ENV === "development";
};

// src/environment/is-firefox/is-firefox.ts
var isFirefox = () => !!navigator.userAgent.toLowerCase().match(/firefox./);

// src/environment/is-internet-explorer/is-internet-explorer.ts
var isInternetExplorer = () => !!navigator.userAgent.toLowerCase().match(/trident./);

// src/environment/is-lighthouse/is-lighthouse.ts
var isLighthouse = (userAgent) => {
  if (!userAgent && isClient()) {
    userAgent = window.navigator.userAgent;
  }
  return ["Lighthouse", "Page Speed", "PageSpeed"].some((part) => (userAgent ?? "").includes(part));
};

// src/environment/is-mobile/is-mobile.ts
var isMobile = () => /android|blackberry|iemobile|ipad|iphone|ipod|opera mini|webos/i.test(navigator.userAgent);

// src/environment/is-motionless/is-motionless.ts
var isMotionless = () => window.matchMedia("(prefers-reduced-motion: reduce)").matches;

// src/environment/is-osx/is-osx.ts
var isOSX = () => {
  const platform = navigator.userAgentData?.platform || navigator.platform || "";
  return !!platform.match(/(ipad|iphone|ipod|mac)/i);
};

// src/environment/is-prod/is-prod.ts
var isProd = () => {
  if (typeof process === "undefined")
    return null;
  return process.env.NODE_ENV === "production";
};

// src/environment/is-server/is-server.ts
var isServer = () => typeof window === "undefined";

// src/environment/is-test/is-test.ts
var isTest = () => process.env.NODE_ENV === "test";

// src/environment/is-touch/is-touch.ts
var isTouch = () => "ontouchstart" in window || !!navigator?.msMaxTouchPoints;

// src/environment/supports-passive/supports-passive.ts
var supportsPassive = () => {
  let supports = false;
  const options = Object.defineProperty({}, "passive", {
    get() {
      supports = true;
      return null;
    }
  });
  try {
    const noop = () => {
    };
    window.addEventListener("test:passive", noop, options);
    window.removeEventListener("test:passive", noop, options);
  } catch {
  }
  return supports;
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  isClient,
  isDev,
  isFirefox,
  isInternetExplorer,
  isLighthouse,
  isMobile,
  isMotionless,
  isOSX,
  isProd,
  isServer,
  isTest,
  isTouch,
  supportsPassive
});
